srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+	/// `beneficiary`.
+	///
+	/// As much funds up to `amount` will be moved as possible. If this is less than `amount`, then
+	/// the `remaining` would be returned, else `Zero::zero()`.
+	pub fn repatriate_reserved(
+		asset_id: &T::AssetId,
+		who: &T::AccountId,
+		beneficiary: &T::AccountId,
+		amount: T::Balance,
+	) -> T::Balance {
+		let b = Self::reserved_balance(asset_id, who);
+		let slash = rstd::cmp::min(b, amount);
+
+		let original_free_balance = Self::free_balance(asset_id, beneficiary);
+		let new_free_balance = original_free_balance + slash;
+		Self::set_free_balance(asset_id, beneficiary, new_free_balance);
+
+		let new_reserve_balance = b - slash;
+		Self::set_reserved_balance(asset_id, who, new_reserve_balance);
+		amount - slash
+	}
+
+	/// Check permission to perform burn, mint or update.
+	///
+	/// # Arguments
+	/// * `asset_id`:  A `T::AssetId` type that contains the `asset_id`, which has the permission embedded.
+	/// * `who`: A `T::AccountId` type that contains the `account_id` for which to check permissions.
+	/// * `what`: The permission to check.
+	///
+	pub fn check_permission(asset_id: &T::AssetId, who: &T::AccountId, what: &PermissionType) -> bool {
+		let permission_versions: PermissionVersions<T::AccountId> = Self::get_permission(asset_id);
+		let permission = permission_versions.into();
+
+		match (what, permission) {
+			(
+				PermissionType::Burn,
+				PermissionLatest {
+					burn: Owner::Address(account),
+					..
+				},
+			) => account == *who,
+			(
+				PermissionType::Mint,
+				PermissionLatest {
+					mint: Owner::Address(account),
+					..
+				},
+			) => account == *who,
+			(
+				PermissionType::Update,
+				PermissionLatest {
+					update: Owner::Address(account),
+					..
+				},
+			) => account == *who,
+			_ => false,
+		}
+	}
+
+	/// Return `Ok` iff the account is able to make a withdrawal of the given amount
+	/// for the given reason.
+	///
+	/// `Err(...)` with the reason why not otherwise.
+	pub fn ensure_can_withdraw(
+		asset_id: &T::AssetId,
+		who: &T::AccountId,
+		_amount: T::Balance,
+		reason: WithdrawReason,
+		new_balance: T::Balance,
+	) -> Result {
+		if asset_id != &Self::staking_asset_id() {
+			return Ok(());
+		}
+
+		let locks = Self::locks(who);
+		if locks.is_empty() {
+			return Ok(());
+		}
+		let now = <system::Module<T>>::block_number();
+		if Self::locks(who)
+			.into_iter()
+			.all(|l| now >= l.until || new_balance >= l.amount || !l.reasons.contains(reason))
+		{
+			Ok(())
+		} else {
+			Err("account liquidity restrictions prevent withdrawal")
+		}
+	}
+
+	// PRIVATE MUTABLES
+
+	/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
+	/// the caller will do this.
+	fn set_reserved_balance(asset_id: &T::AssetId, who: &T::AccountId, balance: T::Balance) {
+		<ReservedBalance<T>>::insert(asset_id, who, balance);
+	}
+
+	/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
```

Comment:
same warning should be given to the above functions that don't manage the imbalance.